### PR TITLE
Support `Eventstats` in PPL

### LINKED ISSUE #660: [FEATURE]Extend ppl `stats` command  functionality
**High level Review**

The OpenSearch Piped Processing Language (PPL) currently lacks some advanced statistical aggregation capabilities similar to those provided by the `eventstats` command in Splunk Search Processing Language (SPL).
This feature request proposes adding new functions and syntax to PPL to enable statistical calculations and aggregations on event data.

Proposed Functionality:

1. **Aggregate statistical calculations:**
   - Calculate **common statistical** measures like `sum`, `count`, `min`, `max`, `avg`, etc., on specific fields or expressions.
   - Support **grouping events** by one or more fields and performing statistical calculations within each group.
   - Allow renaming the calculated fields with custom names.

2. **Conditional aggregations:**
   - Perform statistical calculations based on conditional expressions or filters.
   - Evaluate conditional expressions for each event and aggregate the results (e.g., sum of a conditional expression).

3. **Chaining and nesting:**
   - Enable chaining and nesting of statistical calculations, similar to how `eventstats` commands can be chained in SPL.
   - Allow performing multiple levels of aggregations and calculations in a single query.

4. **Integration with existing PPL syntax:**
   - Seamlessly integrate the new statistical aggregation capabilities with the existing PPL syntax and functions.
   - Ensure compatibility with other PPL features and maintain the overall usability and readability of the language.

**Examples:**

1. Calculate the sum of a conditional expression grouped by a field:

```sql
stats sum(if(field1 = "value" and field2 like "%pattern%", 1, 0)) as conditional_sum by group_field
```

2. Calculate minimum and maximum values of a field grouped by another field:

```sql
stats min(latency_field) as min_latency, max(latency_field) as max_latency by operation_id
```

3. Chain multiple statistical calculations:

```sql
stats sum(count) as total_count by client_id | stats sum(total_count) as overall_total
```

### PR DESCRIPTION
**Pull request by LantaoJin**
### Description
The `eventstats` command enriches your event data with calculated summary statistics. It operates by analyzing specified fields within your events, computing various statistical measures, and then appending these results as new fields to each original event.

Key aspects of `eventstats`:

1. It performs calculations across the entire result set or within defined groups.
2. The original events remain intact, with new fields added to contain the statistical results.
3. The command is particularly useful for comparative analysis, identifying outliers, or providing additional context to individual events.

### Difference between [`stats`](ppl-stats-command.md) and `eventstats`
The `stats` and `eventstats` commands are both used for calculating statistics, but they have some key differences in how they operate and what they produce:

- Output Format:
  - `stats`: Produces a summary table with only the calculated statistics.
  - `eventstats`: Adds the calculated statistics as new fields to the existing events, preserving the original data.
- Event Retention:
  - `stats`: Reduces the result set to only the statistical summary, discarding individual events.
  - `eventstats`: Retains all original events and adds new fields with the calculated statistics.
- Use Cases:
  - `stats`: Best for creating summary reports or dashboards. Often used as a final command to summarize results.
  - `eventstats`: Useful when you need to enrich events with statistical context for further analysis or filtering. Can be used mid-search to add statistics that can be used in subsequent commands.

### Syntax
```sql
eventstats <aggregation>... [by-clause]
```
(check "docs/ppl-lang/ppl-eventstats-command.md" for details)

#### **Event Aggregations**
[See additional command details](ppl-eventstats-command.md)

- `source = table | eventstats avg(a) `
- `source = table | where a < 50 | eventstats avg(c) `
- `source = table | eventstats max(c) by b`
- `source = table | eventstats count(c) by b | head 5`
- `source = table | eventstats stddev_samp(c)`
- `source = table | eventstats stddev_pop(c)`
- `source = table | eventstats percentile(c, 90)`
- `source = table | eventstats percentile_approx(c, 99)`

**Limitation: distinct aggregation could not used in `eventstats`:**_
- `source = table | eventstats distinct_count(c)` (throw exception)

**Aggregations With Span**
- `source = table  | eventstats count(a) by span(a, 10) as a_span`
- `source = table  | eventstats sum(age) by span(age, 5) as age_span | head 2`
- `source = table  | eventstats avg(age) by span(age, 20) as age_span, country  | sort - age_span |  head 2`

**Aggregations With TimeWindow Span (tumble windowing function)**
- `source = table | eventstats sum(productsAmount) by span(transactionDate, 1d) as age_date | sort age_date`
- `source = table | eventstats sum(productsAmount) by span(transactionDate, 1w) as age_date, productId`

**Aggregations Group by Multiple Times**
- `source = table | eventstats avg(age) as avg_state_age by country, state | eventstats avg(avg_state_age) as avg_country_age by country`
- `source = table | eventstats avg(age) as avg_city_age by country, state, city | eval new_avg_city_age = avg_city_age - 1 | eventstats avg(new_avg_city_age) as avg_state_age by country, state | where avg_state_age > 18 | eventstats avg(avg_state_age) as avg_adult_country_age by country`

### Related Issues
Resolves #660 

### Check List
- [x] Updated documentation (docs/ppl-lang/README.md)
- [x] Implemented unit tests
- [x] Implemented tests for combination with other commands
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/ppl-lang/PPL-Example-Commands.md [hunk 1]
```diff
@@ -167,6 +167,34 @@ source = table |  where ispresent(a) |
...
+- `source = table | eventstats avg(a) `
+- `source = table | where a < 50 | eventstats avg(c) `
+- `source = table | eventstats max(c) by b`
+- `source = table | eventstats count(c) by b | head 5`
+- `source = table | eventstats stddev_samp(c)`
+- `source = table | eventstats stddev_pop(c)`
+- `source = table | eventstats percentile(c, 90)`
+- `source = table | eventstats percentile_approx(c, 99)`
+
+**Limitation: distinct aggregation could not used in `eventstats`:**_
```
Comment (YANG-DB) [inline docs/ppl-lang/PPL-Example-Commands.md:182]: please also add it cant be used in conjunction with `stats`  - probably obvious but still need to be noted...

File: docs/ppl-lang/ppl-eventstats-command.md
Comment (YANG-DB) [inline docs/ppl-lang/ppl-eventstats-command.md:1]: this is an excellent document - can u plz create a similar doc for the `stats` command ?
in a new PR ...
thanks!!

File: docs/ppl-lang/PPL-Example-Commands.md (hunk 1 above)
Comment (LantaoJin) [reply docs/ppl-lang/PPL-Example-Commands.md:182]: This limitation is only for the new command `eventstats`, not for `stats`. If we add a limitation note for conjunction with `stats`, similar, a limitation note for conjunction with every other commands would be considered. Would it be gilding the lily?

File: docs/ppl-lang/ppl-eventstats-command.md
Comment (LantaoJin) [reply docs/ppl-lang/ppl-eventstats-command.md:1]: `Stats` command already had one https://github.com/opensearch-project/opensearch-spark/blob/main/docs/ppl-lang/ppl-stats-command.md

File: docs/ppl-lang/PPL-Example-Commands.md (hunk 1 above)
Comment (YANG-DB) [reply docs/ppl-lang/PPL-Example-Commands.md:182]: ok make sense...

**Review by YANG-DB — APPROVED** [review]

File: docs/ppl-lang/ppl-eventstats-command.md [hunk 2]
```diff
@@ -0,0 +1,327 @@
+## PPL `eventstats` command
+
+### Description
+The `eventstats` command enriches your event data with calculated summary statistics. It operates by analyzing specified fields within your events, computing various statistical measures, and then appending these results as new fields to each original event.
+
+Key aspects of `eventstats`:
+
+1. It performs calculations across the entire result set or within defined groups.
+2. The original events remain intact, with new fields added to contain the statistical results.
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply docs/ppl-lang/ppl-eventstats-command.md:9]: > in some cases I would only want to see a subset of fields with the enriched aggregation - should we allow adding the `fields` command after ?

Yes. This command only enriches the row with a new column, it allows to add `fields` command (any other type of commands) after it with `|` symbol.

**Pull request merged**